### PR TITLE
Test files for mknod/fchdir/poll

### PR DESCRIPTION
**Pull request by kuzeyardabulut**
## Description
This PR introduces three new test cases for the mknod, fchdir, and poll syscalls. The test files have been created and tested for validity. They will help us confirm the correct operation of these syscalls.

### fchdir.c
A new C source file, `fchdir.c`, tests the `fchdir` system call functionality. This file opens a directory, changes the current working directory to the newly opened directory, and then verifies this operation by getting the current working directory again.

### mknod.c
We've introduced a C source file, `mknod.c`, to verify the `mknod` system call. This test creates and then removes a character device.

### poll.c
Lastly, a C source file, `poll.c`, has been added to evaluate the `poll` system call. This test script observes if it succeeds using the poll syscall.

### Type of change
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
To reproduce the tests, follow these steps for each system call:

### Test of mknod
- Test of `mknod`- `lind_project/tests/test_cases/mknod.c`
1. Compile `mknod.c` in `lindsh` with `compile tests/test_cases/mknod.c -o mknod.nexe`
2. Copy test file to Lind file system with `lindfs cp /home/lind/lind_project/mknod.nexe mknod.nexe`
3. If all operations were successful, your file should have been copied. You can check the file with `lindfs ls /` command.
4. Run the test file with `lind /mknod.nexe`

### Test of fchdir
- Test of `fchdir` - `lind_project/tests/test_cases/fchdir.c`
1. Compile `fchdir.c` in `lindsh` with `compile tests/test_cases/fchdir.c -o fchdir.nexe`
2. Copy test file to Lind file system with `lindfs cp /home/lind/lind_project/fchdir.nexe fchdir.nexe`
3. To perform the test, create a directory with the `lindfs mkdir automated_tests` command.
4. If all operations are successful, your file should be copied and your directory should be created. You can check the file with `lindfs ls /` command.
5. Run the test file with `lind /fchdir.nexe`

### Test of poll
- Test of `poll` - `lind_project/tests/test_cases/poll.c`
1. Compile `poll.c` in `lindsh` with `compile tests/test_cases/poll.c -o poll.nexe`
2. Copy test file to Lind file system with `lindfs cp /home/lind/lind_project/poll.nexe poll.nexe`
3. If all operations were successful, your file should have been copied. You can check the file with `lindfs ls /` command.
4. Run the test file with `lind /poll.nexe`

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, safeposix-rust)

### CODE REVIEW

**Comment by yizhuoliang** [comment]
Looks good, some minor concerns / suggestions here, all simple
1. `fchdir`: Maybe also try passing it an invalid fd, and see if it correctly return -1 without crashing or anything
2. `mknod`: I'm not sure if any user level code should be priviledged to create character devices
3.  `poll`: Maybe also add a blocked fd to the `fds` array and see if `poll` returns #available fd correctly

File: tests/test_cases/fchdir.c [hunk 1]
```diff
@@ -0,0 +1,59 @@
+#define _GNU_SOURCE
```
Comment (kpan2034) [inline tests/test_cases/fchdir.c:1]: Shouldn't there be a `#undef _GNU_SOURCE` here?

**Review by kpan2034 — CHANGES_REQUESTED** [review]
One change needed in `fchdir.c`.

Also, the build check seems to be failing due to `mknod` not being implemented; I'm not sure if we need to hold off on merging this till the PR in rustposix is merged.

**Review by Yaxuan-w — COMMENTED** [review]
For `poll.c`, should we also test for read (`POLLIN`) and err (`POLLERR`) event type as well?
`mknod.c` which tests for character device file and `fchdir.c` which tests for changing working dictionary by comparing path results look good to me

**Review by kpan2034 — APPROVED** [review]
I see you've made the necessary changes; this looks good to me.

File: tests/test_cases/fchdir.c [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        perror("getcwd() error");
+        return EXIT_FAILURE;
+    }
+    
+    printf("current working directory is: %s :: %s\n", path, result);
+    fflush(stdout);
+    
+
+    // Open a directory
+    int fd = open("automated_tests/", O_RDWR);
```
Comment (RusherRG) [inline tests/test_cases/fchdir.c:27]: Just make sure to append `automated_tests/` to the current working directory path i.e. `result` variable and then open the directory.

File: tests/test_cases/mknod.c [hunk 3]
```diff
@@ -0,0 +1,35 @@
+#undef _GNU_SOURCE
+#define _GNU_SOURCE
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#define DEVICE_PATH "/dev/test" // Defining the path for the device to be created
```
Comment (RusherRG) [inline tests/test_cases/mknod.c:10]: `/dev/test` doesn't exist in lindfs so maybe make use of one of the following:

- /dev/null 
- /dev/urandom 
- /dev/random 
- /dev/zero

Or just create a new file in general and use that.

**Review by RusherRG — CHANGES_REQUESTED** [review]

**Comment by rennergade** [comment]
@kuzeyardabulut 

it looks like mknod is still failing. Looking at your test now it seems native isnt printing anything. I think thats because it doesn't have the appropriate permissions. When you compile the mknod test natively does it run?

Also looking at the fchdir test it seems like it could be moved to dettests, is that not true?

**Comment by RusherRG** [comment]
> Also looking at the fchdir test it seems like it could be moved to dettests, is that not true?

@rennergade I think we can't do that because as seen from the [outputs](https://github.com/Lind-Project/lind_project/actions/runs/5536766485/jobs/10104853710#step:9:1221) of both the tests the current working directory is `/` for lind while it is `/home/lind/lind_project/actions-runner/_work/lind_project/lind_project/tests/test_cases` for the native output. So we would need it to be non-deterministic to compare the change in working directory. 

Also, with respect to the mknod test, could you(@kuzeyardabulut ) try using a path like `getcwd() + /dev/test` or any other path that can be accessed in lind as well as native as we had discussed on Slack.

**Comment by kuzeyardabulut** [comment]
> > Also looking at the fchdir test it seems like it could be moved to dettests, is that not true?
> 
> @rennergade I think we can't do that because as seen from the [outputs](https://github.com/Lind-Project/lind_project/actions/runs/5536766485/jobs/10104853710#step:9:1221) of both the tests the current working directory is `/` for lind while it is `/home/lind/lind_project/actions-runner/_work/lind_project/lind_project/tests/test_cases` for the native output. So we would need it to be non-deterministic to compare the change in working directory.
> 
> Also, with respect to the mknod test, could you(@kuzeyardabulut ) try using a path like `getcwd() + /dev/test` or any other path that can be accessed in lind as well as native as we had discussed on Slack.

Unfortunately, it appears that we are unable to establish this device within any directory where our standard user possesses write permission, for instance, the "/tmp/" and "/home/lind/" directories.

So we decided to not link mknod.

**Comment by rennergade** [comment]
Ok this is good now. Sorry this was such a headache. Great job!

**Pull request merged**